extensions/amp-ad-network-base/0.1/amp-ad-network-base.js

```diff
@@ -0,0 +1,233 @@
+import {sendXhrRequest} from './amp-ad-utils';
+
+const TAG = 'amp-ad-network-base';
+
+/**
+ * @abstract
+ */
+export class AmpAdNetworkBase extends AMP.BaseElement {
+
+  constructor(element) {
+    super(element);
+
+    /** @private {?Promise<!../../../src/service/xhr-impl.FetchResponse>} */
+    this.adPromise_ = null;
+
+    /** @private {Object<string, !./amp-ad-type-defs.Validator>} */
+    this.validators_ = map();
+
+    /** @private {Object<string, !./amp-ad-type-defs.Renderer>} */
+    this.renderers_ = map();
+
+    /** @private {Object<string, string>} */
+    this.recoveryModes_ = map();
+
+    /** @private {?./amp-ad-type-defs.LayoutInfoDef} */
+    this.initialSize_ = null;
+
+    /** @const @private {!Object} */
+    this.context_ = {};
+
+    // Register default error modes.
+    for (const failureType in FailureType) {
+      this.recoveryModes_[failureType] = RecoveryModeType.COLLAPSE;
+    }
+
+    /**
+     * Number of times to retry a failed request. Zero by default.
+     * @type {number}
+     */
+    this.retryLimit_ = 0;
+  }
+
+  /** @override */
+  buildCallback() {
+    this.initialSize_ = {
+      // TODO(levitzky) handle non-numeric values.
+      width: this.element.getAttribute('width'),
+      height: this.element.getAttribute('height'),
+    };
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /** @override */
+  layoutCallback() {
+    dev().assert(this.adPromise_, 'layoutCallback invoked before XHR request!');
+    return this.adPromise_
+        .then(response => this.invokeValidator_(response))
+        .then(validatorResult => this.invokeRenderer_(validatorResult))
+        .catch(error => this.handleFailure_(error.type, error.msg));
+  }
+
+  /** @override */
+  onLayoutMeasure() {
+    this.sendRequest_();
+  }
+
+
+  /**
+   * @param {!./amp-ad-type-defs.FailureType} failure
+   * @param {!./amp-ad-type-defs.RecoveryModeType} recovery
+   * @final
+   */
+  onFailure(failure, recovery) {
+    if (this.recoveryModes_[failure]) {
+      dev().warn(TAG,
+          `Recovery mode for failure type ${failure} already registered!`);
+    }
+    this.recoveryModes_[failure] = recovery;
+  }
+
+  /**
+   * @param {!./amp-ad-type-defs.Validator} validator
+   * @param {string=} type
+   * @final
+   */
+  registerValidator(validator, type = 'default') {
+    if (this.validators_[type]) {
+      dev().warn(TAG, `${type} validator already registered.`);
+    }
+    this.validators_[type] = validator;
+  }
+
+  /**
+   * @param {!./amp-ad-type-defs.Renderer} renderer
+   * @param {string} type
+   * @final
+   */
+  registerRenderer(renderer, type) {
+    if (this.renderers_[type]) {
+      dev().warn(TAG, `Rendering mode already registered for type '${type}'`);
+    }
+    this.renderers_[type] = renderer;
+  }
+
+  /**
+   * @param {string} name
+   * @param {*} value
+   */
+  setContextField(name, value) {
```

Comment:
I guess we don't need this. Because we are passing the context all way along the flow

Reply:
This provides subclasses with ways of adding arbitrary data to the context, unless we make this.context public, but this approach feels more inline with OOP.

Reply:
OK, that's fine.

You might just have 
```
getContext() {...};
```

otherwise, you'll need a getter method

Reply:
Done.